### PR TITLE
refine the code and merge to sycl-develop

### PR DESCRIPTION
**Pull request by jiyang1011**
features list:
1. block 2d load encapsulation
    combination : block_y X block_x X (8 x 16) where block_y in {1, 2, 4} and block_x in {1, 2, 4}
and structure name follows the rule `data_type x {8x16} x {block_y} x {block_x}_{V, N}` and `N` means `not-transposed` ,  `V` means `VNNI-packed`
2. include some sycl type
3. include openCL extension intrinsic

### CODE REVIEW

**Review by taozha2 — APPROVED** [review]

**Comment by AD2605** [comment]
Could you please provide a description in the PR, listing the features being added ? Thanks.

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # PyCache files
 __pycache__/
-cutlass_library.egg-info/
\ No newline at end of file
+cutlass_library.egg-info/
+build/
+*.patch
```
Comment (AD2605) [inline .gitignore:5]: Rather than adding extra folders in the .gitignore like build and for patch files, a better approach would be to include these in your global .gitignore config, this way we can keep the changes minimal.

File: include/cute/arch/copy_xe.hpp [hunk 2]
```diff
@@ -1,45 +1,201 @@
+/***************************************************************************************************
+ * Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:2]: Should the copyright be Nvidia's for new files ?

File: include/cute/arch/copy_xe.hpp [hunk 3]
```diff
@@ -1,45 +1,201 @@
...
-    }
+struct XE_2D_U16X8X16X1X1_LD_N
+{
+  template <class T>
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    *(ushort8 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
+        (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:86]: Should we have error checking for invalid pipeline creation, similar to how Nvidia has done for their various architectures, i.e.
```
#if defined(ARCH_PVC)
*(ushort8 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
        (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
#else
CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
#endif
```

Comment (AD2605) [reply include/cute/arch/copy_xe.hpp:86]: Similarly for all other structs as well

File: include/cute/arch/copy_xe.hpp [hunk 4]
```diff
@@ -1,45 +1,201 @@
...
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord, uint8 data));
+SYCL_DEVICE_BUILTIN(ushort8 __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+SYCL_DEVICE_BUILTIN(uint8 __builtin_IB_subgroup_block_read_flat_u32_m8k16v1(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+
+/// Load A
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:57]: please remove this comment, as from what I understand this file just lists the different functions to be used for copy, and not anything specific to the "A" matrix

File: include/cute/arch/copy_xe.hpp [hunk 5]
```diff
@@ -1,45 +1,201 @@
...
-    template<class T>
-    CUTE_HOST_DEVICE static void copy(void* baseoffset, int width, int height, int pitch, int2_ coord, const T* src)
-    {
-        if constexpr(sizeof(T)==sizeof(uint)) {
-            __builtin_IB_subgroup_block_write_flat_u32_m8k16v1((long)baseoffset, width - 1, height - 1, pitch - 1, coord, *(uint8*)src);
-        } else {
-            static_assert(false);
-        }
-    }
+struct XE_2D_U16X8X16X1X1_LD_N
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:78]: small `x` rather than the capitalized version `X`, i.e.
it should be 

```
XE_2D_U16x8x16x1x1_LD_N
```

Comment (AD2605) [reply include/cute/arch/copy_xe.hpp:78]: Similarly for all other structs as well

File: include/cute/arch/copy_xe.hpp (hunk 2 above)
Comment (AD2605) [reply include/cute/arch/copy_xe.hpp:2]: Same comment for all other new files as well

File: include/cute/arch/copy_xe.hpp [hunk 6]
```diff
@@ -1,45 +1,201 @@
...
+{
+  template <class T>
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:81]: The macro `CUTE_HOST_DEVICE` expands to `__forceinline__ __host__ __device__`, in which `__host__ __device__` are very cuda specific function specifiers, hence I suppose we do not need them here ?

Comment (AD2605) [reply include/cute/arch/copy_xe.hpp:81]: Same goes for all other structs as well

File: include/cute/util/print.hpp [hunk 7]
```diff
@@ -34,6 +34,8 @@
 
 #include <cute/util/type_traits.hpp>
 
+#define printf sycl::ext::oneapi::experimental::printf
```
Comment (AD2605) [inline include/cute/util/print.hpp:37]:
```suggestion
#if defined(CUTLASS_ENABLE_SYCL)
#define printf sycl::ext::oneapi::experimental::printf
#endif
```

**Review by AD2605 — CHANGES_REQUESTED** [review]

File: include/cute/arch/copy_xe.hpp (hunk 6 above)
Comment (AD2605) [reply include/cute/arch/copy_xe.hpp:81]: Nevermind, I found `CUDA_HOST_DEVICE` is defined as 
```
__attribute__((always_inline))
``` 
for the sycl path, closing this comment

File: include/cutlass/bfloat16.h [hunk 8]
```diff
@@ -119,7 +119,7 @@ struct alignas(2) bfloat16_t {
     asm("cvt.rn.bf16.f32 %0, %1;\n" : "=h"(storage) : "f"(x));
 
     #elif defined(CUTLASS_ENABLE_SYCL)
-
+
```
Comment (mehdi-goli) [inline include/cutlass/bfloat16.h:122]: Nit-pick: here is the noise in the pr.

File: include/cute/atom/copy_traits_xe.hpp [hunk 9]
```diff
@@ -1,80 +1,266 @@
...
-        }
-    };
-
-    template <class Copy, class GEngine, class GLayout>
-    auto make_xe_2d_copy(Tensor<GEngine, GLayout> gtensor)
+        static_assert(is_rmem<TS>::value);
+        int H = size<0>(traits.tensor);
+        int W = size<1>(traits.tensor) * sizeof(typename decltype(traits.tensor)::engine_type::value_type);
+        auto [y, x] = dst.data().coord_;
+        CopyOp::copy(traits.tensor.data().get(), W, H, W, int2_{x, y}, &*src.data());
```
Comment (muhammad-tanvir-1211) [inline include/cute/atom/copy_traits_xe.hpp:225]: We should cast these coordinates to `int` in order to avoid the compile time error below. Note that this error arises when calling copy from the cutlass PVC pipeline and not through the cute PVC sample.
```suggestion
        CopyOp::copy(traits.tensor.data().get(), W, H, W, int2_{static_cast<int>(x), static_cast<int>(y)}, &*src.data());
```

`
error: non-constant-expression cannot be narrowed from type 'std::tuple_element<0, const std::tuple<unsigned long, unsigned long>>::type' (aka 'const unsigned long') to 'int' in initializer list [-Wc++11-narrowing]
   67 |             XE_2D_SAVE::copy(traits.tensor.data(), W, H, W, int2_{x, y}, &*src.data());
      |                                                                      ^
/home/tanvir/work/cutlass/include/cute/atom/copy_traits_xe.hpp:67:70: note: insert an explicit cast to silence this issue
   67 |             XE_2D_SAVE::copy(traits.tensor.data(), W, H, W, int2_{x, y}, &*src.data());
      |                                                                      ^
      |                                                                      static_cast<int>( )
`

File: include/cute/util/sycl_vec.hpp [hunk 10]
```diff
@@ -1,17 +1,55 @@
...
-template<class T, int N> using vector_t = typename sycl::vec<T,N>::vector_t;
-#else 
-template<class T, int N> using vector_t = sycl::vec<T,N>;
+#ifdef __SYCL_DEVICE_ONLY__
+template <class T, int N> using vector_t = typename sycl::vec<T, N>::vector_t;
+#else
+template <class T, int N> using vector_t = sycl::vec<T, N>;
 #endif
 
-using float8 = vector_t<float, 8>;
```
Comment (muhammad-tanvir-1211) [inline include/cute/util/sycl_vec.hpp:12]: Isn't `float8` required for the `mma` operation?

File: include/cute/util/sycl_vec.hpp [hunk 11]
```diff
@@ -1,17 +1,55 @@
...
-using ushort8 = vector_t<ushort, 8>;
-using int2_ = vector_t<int, 2>; //conflicts with vector_types
-using int8 = vector_t<int, 8>;
-using uint8 = vector_t<uint, 8>;
+// using float8 = vector_t<float, 8>;
+// using short8 = vector_t<short, 8>;
+// using ushort8 = vector_t<ushort, 8>;
+using int2_ = vector_t<int, 2>; // conflicts with vector_types
+// using int8 = vector_t<int, 8>;
+// using uint8 = vector_t<uint, 8>;
```
Comment (muhammad-tanvir-1211) [inline include/cute/util/sycl_vec.hpp:47]: I think you missed removing the comments here.

File: README-sycl.md [hunk 12]
```diff
@@ -52,7 +52,8 @@ ninja [EXAMPLE_NAME]_sycl
 You can run it like this from your build directory
 
 ```
-LD_LIBRARY_PATH=/path/to/sycl/install/lib ./examples/cute/tutorial/[EXAMPLE_NAME]_sycl
+LD_LIBRARY_PATH=/path/to/sycl/install/lib 
+./examples/cute/tutorial/[EXAMPLE_NAME]_sycl
```
Comment (aacostadiaz) [inline README-sycl.md:56]: Could you revert this change?

File: README-sycl.md [hunk 13]
```diff
@@ -62,7 +63,8 @@ LD_LIBRARY_PATH=/path/to/sycl/install/lib ./examples/cute/tutorial/[EXAMPLE_NAME
  ```
  You can run it like this from your build directory
  ```
-  NVIDIA_TF32_OVERRIDE=1 LD_LIBRARY_PATH=/path/to/sycl/install/lib ./examples/14_ampere_tf32_tensorop_gemm/14_ampere_tf32_tensorop_gemm_cute
+  NVIDIA_TF32_OVERRIDE=1 LD_LIBRARY_PATH=/path/to/sycl/install/lib 
+ ./examples/14_ampere_tf32_tensorop_gemm/14_ampere_tf32_tensorop_gemm_cute
```
Comment (aacostadiaz) [inline README-sycl.md:67]: Same here

File: include/cute/atom/copy_traits_xe.hpp [hunk 14]
```diff
@@ -1,80 +1,266 @@
...
+    // using ThrID   = Layout<_16>; //TODO: I think it should be 16 (copy is per subgroup) - but static_assert fails
+    using ThrID = Layout<_1>;
+    using NumBits = Int<sizeof(typename GTensor::engine_type::value_type) * 8>; // hacky: does vec of 8
+    // Map from (src-thr,src-val) to bit
+    using SrcLayout = Layout<Shape<_1, NumBits>>; // TODO:  is _1 correct?
+    // Map from (dst-thr,dst-val) to bit
+    using DstLayout = Layout<Shape<_1, NumBits>>;
+    // Reference map from (thr,val) to bit
+    using RefLayout = SrcLayout;
+    using CopyInternalType = ushort;
```
Comment (aacostadiaz) [inline include/cute/atom/copy_traits_xe.hpp:72]: Instead of adding this, you can pass the `GTensor` to `XE_2D_LD_Unpack<XE_2D_U16X8X16X1X1_LD_N, GTensor>` and get the type using `typename GTensor::engine_type::value_type)`

File: include/cute/atom/copy_traits_xe.hpp [hunk 15]
```diff
@@ -1,80 +1,266 @@
...
+    }
+};
+
+template <class GTensor>
+struct Copy_Traits<XE_2D_U16X8X16X1X1_LD_N, GTensor>
+     : XE_2D_LD_Unpack<XE_2D_U16X8X16X1X1_LD_N>
+{
+    // using ThrID   = Layout<_16>; //TODO: I think it should be 16 (copy is per subgroup) - but static_assert fails
+    using ThrID = Layout<_1>;
+    using NumBits = Int<sizeof(typename GTensor::engine_type::value_type) * 8>; // hacky: does vec of 8
```
Comment (aacostadiaz) [inline include/cute/atom/copy_traits_xe.hpp:65]: Should this value be part of of copy operation? Something like XE_2D_U16X8X16X1X1_LD_N::ElementCount defined as the number of elements copied.

File: include/cute/util/sycl_vec.hpp [hunk 16]
```diff
@@ -1,17 +1,55 @@
...
-using float8 = vector_t<float, 8>;
-using short8 = vector_t<short, 8>;
-using ushort8 = vector_t<ushort, 8>;
-using int2_ = vector_t<int, 2>; //conflicts with vector_types
-using int8 = vector_t<int, 8>;
-using uint8 = vector_t<uint, 8>;
+// using float8 = vector_t<float, 8>;
+// using short8 = vector_t<short, 8>;
+// using ushort8 = vector_t<ushort, 8>;
+using int2_ = vector_t<int, 2>; // conflicts with vector_types
```
Comment (aacostadiaz) [inline include/cute/util/sycl_vec.hpp:45]: Does it work if we use the int2 type defined in Cutlass?

File: include/cute/util/sycl_vec.hpp [hunk 17]
```diff
@@ -1,17 +1,55 @@
...
+// using int8 = vector_t<int, 8>;
+// using uint8 = vector_t<uint, 8>;
+// using ushort16 = vector_t<ushort, 16>;
+// using uint16 = vector_t<uint, 16>;
+
+typedef ushort __attribute__((ext_vector_type(8))) ushort8_t;
+typedef ushort __attribute__((ext_vector_type(16))) ushort16;
+typedef ushort __attribute__((ext_vector_type(32))) ushort32;
+typedef ushort __attribute__((ext_vector_type(64))) ushort64;
+typedef uint __attribute__((ext_vector_type(32))) uint32;
```
Comment (aacostadiaz) [inline include/cute/util/sycl_vec.hpp:55]: Same here, if these values are already defined in cutlass and we can use that instead we could remove this file entirely.

File: include/cute/atom/copy_traits_xe.hpp [hunk 18]
```diff
@@ -1,80 +1,266 @@
...
-            XE_2D_LOAD::copy(traits.tensor.data().get(), W, H, W, int2_{x, y}, &*dst.data());
-        }
-    };
-
-
-    template <class GTensor>
-    struct Copy_Traits<XE_2D_SAVE, GTensor>
+        static_assert(is_rmem<TD>::value);
+        int H = size<0>(traits.tensor);
+        // int W = size<1>(traits.tensor) * sizeof(typename decltype(traits.tensor)::engine_type::value_type);
```
Comment (aacostadiaz) [inline include/cute/atom/copy_traits_xe.hpp:52]: Could you remove all the unnecessary comments in this file? Like commented LoC, TODOs and hacky: comments

**Review by aacostadiaz — CHANGES_REQUESTED** [review]
Thank you for addressing my previous comments. The PR looks much better now. I've left some other comments.

File: include/cute/util/sycl_vec.hpp (hunk 16 above)
Comment (jiyang1011) [reply include/cute/util/sycl_vec.hpp:45]: we ever hacked this type to avoid the compiling error and nv int2 seems not ideal.

File: include/cute/util/sycl_vec.hpp (hunk 17 above)
Comment (jiyang1011) [reply include/cute/util/sycl_vec.hpp:55]: we use openCL type to meet the openCL extension intrinsic, which is a hacky now

File: include/cute/atom/copy_traits_xe.hpp (hunk 14 above)
Comment (jiyang1011) [reply include/cute/atom/copy_traits_xe.hpp:72]: the` CopyInternalType ` is not always same to `GTensor::engine_type::value_type`

File: include/cute/util/sycl_vec.hpp [hunk 19]
```diff
@@ -1,17 +1,53 @@
+/***************************************************************************************************
+ * Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (muhammad-tanvir-1211) [inline include/cute/util/sycl_vec.hpp:2]: Thank you for addressing my previous comments. I think you missed this change in the license here.
```suggestion
 * Copyright (c) 2024 - 2024 Codeplay Software Ltd. All rights reserved.
```

File: include/cute/arch/copy_xe.hpp [hunk 20]
```diff
@@ -1,45 +1,240 @@
...
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    *(ushort8 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
+        (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:89]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
      *(ushort8 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
          (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

A bit nitpicky, hence providing the suggestions as then you can commit them quickly, the indent in the `#ifdef` blocks is missing. 

Similarly for all `structs`

File: include/cute/arch/copy_xe.hpp [hunk 21]
```diff
@@ -1,45 +1,240 @@
...
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 4, "Expected T to have size 4");
+    *(uint8 *)dst = __builtin_IB_subgroup_block_read_flat_u32_m8k16v1(
+          (long)baseoffset, width - 1, height - 1, pitch - 1, coord); 
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:105]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 4, "Expected T to have size 4");
      *(uint8 *)dst = __builtin_IB_subgroup_block_read_flat_u32_m8k16v1(
            (long)baseoffset, width - 1, height - 1, pitch - 1, coord); 
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif                                     
```

File: include/cute/arch/copy_xe.hpp [hunk 22]
```diff
@@ -1,45 +1,240 @@
...
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)                                  
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    *(uint8 *)dst = __builtin_IB_subgroup_block_read_flat_u32_m8k16v1(
+          (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:121]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)                                  
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
      *(uint8 *)dst = __builtin_IB_subgroup_block_read_flat_u32_m8k16v1(
            (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif                                      
```

File: include/cute/arch/copy_xe.hpp [hunk 23]
```diff
@@ -1,45 +1,240 @@
...
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    *(ushort64 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m32k16v2(
+        long(baseoffset), width - 1, height - 1, pitch - 1, coord);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:137]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
      *(ushort64 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m32k16v2(
          long(baseoffset), width - 1, height - 1, pitch - 1, coord);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 24]
```diff
@@ -1,45 +1,240 @@
...
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    *(ushort32*) dst = __builtin_IB_subgroup_block_read_flat_u16_m16k16v2(
+        long(baseoffset), width - 1, height - 1, pitch - 1, coord);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:153]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
      *(ushort32*) dst = __builtin_IB_subgroup_block_read_flat_u16_m16k16v2(
          long(baseoffset), width - 1, height - 1, pitch - 1, coord);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 25]
```diff
@@ -1,45 +1,240 @@
...
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    ushort16 tmp = (intel_subgroup_block_read_u16_m8k16v2(
+        (__global void *)baseoffset, width, height, pitch, coord));
+    *(ushort16 *)dst = *reinterpret_cast<ushort16 *>(&tmp);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:170]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
      ushort16 tmp = (intel_subgroup_block_read_u16_m8k16v2(
          (__global void *)baseoffset, width, height, pitch, coord));
      *(ushort16 *)dst = *reinterpret_cast<ushort16 *>(&tmp);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 26]
```diff
@@ -1,45 +1,240 @@
...
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+      *(ushort32*) dst = __builtin_IB_subgroup_block_read_flat_u16_m32k16v1(
+          long(baseoffset), width - 1, height - 1, pitch - 1, coord);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:186]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
        *(ushort32*) dst = __builtin_IB_subgroup_block_read_flat_u16_m32k16v1(
            long(baseoffset), width - 1, height - 1, pitch - 1, coord);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 27]
```diff
@@ -1,45 +1,240 @@
...
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 4, "Expected T to have size 4");
+    uint16 tmp = __builtin_IB_subgroup_block_read_flat_u32_m16k16v1(
+        long(baseoffset), width - 1, height - 1, pitch - 1, coord);
+    *(uint16 *)dst = *reinterpret_cast<uint16 *>(&tmp);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:203]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 4, "Expected T to have size 4");
      uint16 tmp = __builtin_IB_subgroup_block_read_flat_u32_m16k16v1(
          long(baseoffset), width - 1, height - 1, pitch - 1, coord);
      *(uint16 *)dst = *reinterpret_cast<uint16 *>(&tmp);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 28]
```diff
@@ -1,45 +1,240 @@
...
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    uint16 tmp = __builtin_IB_subgroup_block_read_flat_u32_m16k16v1(
+        long(baseoffset), width - 1, height - 1, pitch - 1, coord);
+    *(uint16 *)dst = *reinterpret_cast<uint16 *>(&tmp);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:220]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 2, "Expected T to have size 2");
      uint16 tmp = __builtin_IB_subgroup_block_read_flat_u32_m16k16v1(
          long(baseoffset), width - 1, height - 1, pitch - 1, coord);
      *(uint16 *)dst = *reinterpret_cast<uint16 *>(&tmp);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 29]
```diff
@@ -1,45 +1,240 @@
...
+  CUTE_HOST_DEVICE static void copy(void *baseoffset, int width, int height,
+                                    int pitch, int2_ coord, const T *src) {
+    #if defined(ARCH_PVC_ACTIVATED)
+    static_assert(sizeof(T) == 4, "Expected T to have size 4");
+    __builtin_IB_subgroup_block_write_flat_u32_m8k16v1(
+        (long)baseoffset, width - 1, height - 1, pitch - 1, coord,
+        *(uint8 *)src);
+    #else
+    CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
+    #endif
```
Comment (AD2605) [inline include/cute/arch/copy_xe.hpp:236]:
```suggestion
    #if defined(ARCH_PVC_ACTIVATED)
      static_assert(sizeof(T) == 4, "Expected T to have size 4");
      __builtin_IB_subgroup_block_write_flat_u32_m8k16v1(
          (long)baseoffset, width - 1, height - 1, pitch - 1, coord,
          *(uint8 *)src);
    #else
      CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
    #endif
```

File: include/cute/arch/copy_xe.hpp [hunk 30]
```diff
@@ -1,45 +1,240 @@
...
+#if defined(CUTLASS_ENABLE_SYCL)
+  #define ARCH_PVC_ACTIVATED 1
+#endif
+
 #ifdef __SYCL_DEVICE_ONLY__
 #define SYCL_DEVICE_BUILTIN(x) SYCL_EXTERNAL extern "C" x
 #else
-#define SYCL_DEVICE_BUILTIN(x) \
-    inline x { assert(false); }
+#define SYCL_DEVICE_BUILTIN(x)                                                 \
```
Comment (mehdi-goli) [inline include/cute/arch/copy_xe.hpp:46]:
```suggestion
#define SYCL_DEVICE_BUILTIN(x)  \
```

**Review by AD2605 — APPROVED** [review]

**Review by tdeng5 — APPROVED** [review]

File: include/cute/util/sycl_vec.hpp (hunk 17 above)
Comment (aacostadiaz) [reply include/cute/util/sycl_vec.hpp]: Could you define all the type defs in this file under a `namespace cute::intel` block? We don't want to expose any of this in the global scope.

File: include/cute/util/sycl_vec.hpp (hunk 16 above)
Comment (aacostadiaz) [reply include/cute/util/sycl_vec.hpp]: nv int2 is fine for now. There is a PR open to remove all dependencies on CUDA libraries. So it will end up being a non-nv int2.

Another option is to change the name. Since you are only using it for coordinates, you can call it coord_t.

**Review by aacostadiaz — APPROVED** [review]
LGTM. Thanks for addressing the comments

File: include/cute/atom/copy_traits_xe.hpp [hunk 31]
```diff
@@ -1,80 +1,306 @@
...
+  using RefLayout = DstLayout;
+  using CopyInternalType = ushort;
+};
+
+template <class GTensor>
+struct Copy_Traits<XE_2D_U16x8x16x2x2_LD_N, GTensor>
+     : XE_2D_LD_Unpack<XE_2D_U16x8x16x2x2_LD_N, GTensor>
+{
+  // Logical thread id to thread idx
+  using ThrID = Layout<_16>;
```
Comment (rolandschulz) [inline include/cute/atom/copy_traits_xe.hpp:116]: Would it make sense to move more of this into the base class? Most of them (e.g. `ThrID`) are the same for all.
